identity/aziot-identityd/src/lib.rs

```diff
@@ -702,83 +634,191 @@ impl Api {
         Ok(device)
     }
 
+    async fn issue_local_identity(
+        &self,
+        module_id: &str,
+    ) -> Result<aziot_identity_common::Identity, Error> {
+        let localid = self.settings.localid.as_ref().ok_or_else(|| {
+            Error::Internal(InternalError::BadSettings(std::io::Error::new(
+                std::io::ErrorKind::InvalidInput,
+                "no local id settings specified",
+            )))
+        })?;
+
+        let local_identity =
+            match self
+                .local_identities
+                .get(&aziot_identity_common::ModuleId(module_id.to_owned()))
+            {
+                None => {
+                    return Err(Error::invalid_parameter(
+                        "module_id",
+                        format!("no local identity found for {}", module_id),
+                    ))
+                }
+                Some(opts) => {
+                    let attributes = opts.as_ref().map_or(
+                        aziot_identity_common::LocalIdAttr::default(),
+                        |opts| match opts {
+                            settings::LocalIdOpts::X509 { attributes } => *attributes,
+                        },
+                    );
+
+                    let subject = format!("{}.{}", module_id, localid.domain);
+                    let (cert, key) = self
+                        .create_identity_cert_if_not_exist_or_expired(
+                            KeyType::Raw,
+                            module_id,
+                            subject.as_str(),
+                            Some(attributes),
+                        )
+                        .await?;
+
+                    aziot_identity_common::Identity::Local(aziot_identity_common::LocalIdSpec {
```

Comment:
We need a new spec format with a new type of different auth_type that applies only to LocalIdSpec because you are not always passing the key_handle back now.

Reply:
What should the spec look like? Currently, it's
```
{
   "type":"local",
   "spec":{
      "attr":"client",
      "auth":{
         "type":"x509",
         "keyHandle":"...",
         "certId":"..."
      }
   }
}
```

I'll rename keyHandle and certId to key and cert, since they'll be raw bytes. Any other changes needed?

Reply:
Checking the identity workload API. The expiration is also returned there - [https://github.com/Azure/iotedge/blob/iiot/edgelet/edgelet-http-workload/src/server/cert/mod.rs#L33](https://github.com/Azure/iotedge/blob/iiot/edgelet/edgelet-http-workload/src/server/cert/mod.rs#L33)

---
In reply to: [513769412](https://github.com/Azure/iot-identity-service/pull/61#discussion_r513769412) [](ancestors = 513769412)

Reply:
Changed the spec to add expiration date.
```
{
   "type":"local",
   "spec":{
       "private_key":"...",
       "certificate":"...",
       "expiration":"..."
   }
}
```